DylibBinder/XmlGenerator.cs

```diff
@@ -0,0 +1,258 @@
+﻿using System;
+using System.Linq;
+using System.Xml;
+
+namespace DylibBinder {
+	public class XmlGenerator {
+		public XmlGenerator (DBTopLevel dBTopLevel, string outputPath)
+		{
+			XmlWriter writer = CreateWriter (outputPath);
+			WriteIntro (writer, dBTopLevel);
+			WriteTypeDeclarations (writer, dBTopLevel);
+			CloseWriter (writer);
+		}
+
+		XmlWriter CreateWriter (string outputPath)
+		{
+			XmlWriterSettings settings = new XmlWriterSettings ();
+			settings.Indent = true;
+			settings.IndentChars = "   ";
+			return XmlWriter.Create (outputPath, settings);
+		}
+
+		void WriteIntro (XmlWriter writer, DBTopLevel dBTopLevel) {
+			writer.WriteStartDocument ();
+			writer.WriteStartElement ("xamreflect");
+			writer.WriteAttributeString ("version", dBTopLevel.XmlVersion);
+
+			writer.WriteStartElement ("modulelist");
+
+			writer.WriteStartElement ("module");
+			writer.WriteAttributeString ("name", dBTopLevel.ModuleName);
+			writer.WriteAttributeString ("swiftVersion", dBTopLevel.SwiftVersion);
+		}
+
+
+		void WriteTypeDeclarations (XmlWriter writer, DBTopLevel dBTopLevel)
+		{
+			foreach (var typeDeclaration in dBTopLevel.DBTypeDeclarations.TypeDeclarations) {
+
+				writer.WriteStartElement ("typedeclaration");
+				WriteMultipleAttributeStrings (writer, ("kind", typeDeclaration.Kind),
+											   ("name", typeDeclaration.Name), ("accessibility", typeDeclaration.Accessibility),
+											   ("isDeprecated", typeDeclaration.IsDeprecated), ("isUnavailable", typeDeclaration.IsUnavailable),
+											   ("isObjC", typeDeclaration.IsObjC), ("isFinal", typeDeclaration.IsFinal));
+
+				if (typeDeclaration.Kind == "protocol")
+					WriteProtocolTypeDeclaration (writer, typeDeclaration);
+				else
+					WriteTypeDeclaration (writer, typeDeclaration);
+				writer.WriteEndElement ();
+			}
+		}
+
+		void WriteTypeDeclaration (XmlWriter writer, DBTypeDeclaration typeDeclaration)
+		{
+			WriteMembers (writer, typeDeclaration);
+			WriteInnerTypes (writer, typeDeclaration);
+			WriteGenericParameters (writer, typeDeclaration);
+			WriteAssocTypes (writer, typeDeclaration);
+			WriteElements (writer, typeDeclaration);
```

Comment:
In all the methods, we are passing the writer aroud, is that an instance variable? If it is, do we need to pass it? Else, you are writing instance methods that are not really instance methods, and then we are back to a class with a lot of static methods, right? 

We probably want to make this private instance methods and use the writer in a private variable.

Reply:
Do you mean putting the XmlWriter into a private property?